In [6]:
#loading libraries
import pandas as pd

import warnings
warnings.filterwarnings('ignore')

In [7]:
# loading the file to get the sheet names
xls= pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name= None )
xls.keys()

dict_keys(['Title Sheet', 'Transactions', 'NewCustomerList', 'CustomerDemographic', 'CustomerAddress'])

In [8]:
# loading each sheet to a dataframe
# from view in excel title sheet is not needed
# also the CustomerDemographic sheet was a test run to explain the task
Transactions_df= pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='Transactions', skiprows=(1))
NewCustomerList_df= pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='NewCustomerList', skiprows=(1))
CustomerAddress_df= pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerAddress', skiprows=(1))

# the train dataset for basic checks
CustomerDemographic_df= pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic', skiprows=(1))


## Data Quality Assessment
    This segment contains details and analysis for task 1.

### Assessing the Transaction file

In [9]:
# viewing the column and shape of the dataframe
Transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

- It has 20000 entries and 13 columns
- Only 5 columns do not have null values, so i'll be analyzing these null values
- the id columns are int...as is common with excel files
- the product_first_sold_date should be datetime and not float. From further view...this column is not clear.


In [10]:
Transactions_df.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [11]:
Transactions_df.sample(5)

transaction_id  product_id  customer_id transaction_date  online_order  \
14077           14078           1         2318       2017-04-22           0.0   
17814           17815          85          220       2017-05-22           0.0   
15857           15858           9         2029       2017-09-02           1.0   
14114           14115          25           21       2017-02-12           1.0   
11432           11433          95         2670       2017-11-02           1.0   

      order_status           brand product_line product_class product_size  \
14077     Approved  Giant Bicycles     Standard        medium       medium   
17814     Approved        WeareA2B     Standard        medium       medium   
15857     Approved      OHM Cycles         Road        medium       medium   
14114     Approved  Giant Bicycles         Road        medium       medium   
11432     Approved      OHM Cycles      Touring           low       medium   

       list_price  standard_cost  product_first_sold_date  
14077     1403.50         954.82                  33549.0  
17814     1228.07         400.91                  36668.0  
15857      742.54         667.40                  33549.0  
14114     1538.99         829.65                  36334.0  
11432     1073.07         933.84                  35455.0

- The dataset contains transations for the year 2017 only.

In [12]:
# checking for duplicates
Transactions_df.duplicated().sum()

0

-There are no duplicates in the data.

In [13]:
# checking for null values
Transactions_df.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [14]:
#selecting columns with null values in product_line column
Null_transactions= Transactions_df[Transactions_df['product_line'].isna()]
Null_transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
136             137           0          431       2017-09-23           0.0   
159             160           0         3300       2017-08-27           0.0   
366             367           0         1614       2017-03-10           0.0   
406             407           0         2559       2017-06-14           1.0   
676             677           0         2609       2017-07-02           0.0   

    order_status brand product_line product_class product_size  list_price  \
136     Approved   NaN          NaN           NaN          NaN     1942.61   
159     Approved   NaN          NaN           NaN          NaN     1656.86   
366     Approved   NaN          NaN           NaN          NaN      850.89   
406     Approved   NaN          NaN           NaN          NaN      710.59   
676     Approved   NaN          NaN           NaN          NaN     1972.01   

     standard_cost  product_first_sold_date  
136            NaN                      NaN  
159            NaN                      NaN  
366            NaN                      NaN  
406            NaN                      NaN  
676            NaN                      NaN

In [15]:
Null_transactions.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 197 entries, 136 to 19871
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           197 non-null    int64         
 1   product_id               197 non-null    int64         
 2   customer_id              197 non-null    int64         
 3   transaction_date         197 non-null    datetime64[ns]
 4   online_order             195 non-null    float64       
 5   order_status             197 non-null    object        
 6   brand                    0 non-null      object        
 7   product_line             0 non-null      object        
 8   product_class            0 non-null      object        
 9   product_size             0 non-null      object        
 10  list_price               197 non-null    float64       
 11  standard_cost            0 non-null      float64       
 12  product_first_sold_date  0 non-n

- All the values in the null_transactions_df have no product_id 
- these columns should be dropped: this will take care of null values in 5 other columns as these are cancelled or uncompleted transactions 


In [16]:
Transactions_df.dropna(subset = ['product_line'], inplace=True)

In [17]:
Transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19803 non-null  int64         
 1   product_id               19803 non-null  int64         
 2   customer_id              19803 non-null  int64         
 3   transaction_date         19803 non-null  datetime64[ns]
 4   online_order             19445 non-null  float64       
 5   order_status             19803 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               19803 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

- now i'm left with only one column with na values, the online_order column

In [18]:
Transactions_df.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               358
order_status                 0
brand                        0
product_line                 0
product_class                0
product_size                 0
list_price                   0
standard_cost                0
product_first_sold_date      0
dtype: int64

-358 is still a lot of null values and this should be fixed wile gathering the data instead.
- this can be fixed y dropping the values or filling it with one of the two options...after clarifying which it most likely is

In [19]:
#dropping null values in the last column
#This can be fixed by replacing it with either online or offline
Transactions_df.dropna(subset=['online_order'], inplace= True)
Transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19445 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19445 non-null  int64         
 1   product_id               19445 non-null  int64         
 2   customer_id              19445 non-null  int64         
 3   transaction_date         19445 non-null  datetime64[ns]
 4   online_order             19445 non-null  float64       
 5   order_status             19445 non-null  object        
 6   brand                    19445 non-null  object        
 7   product_line             19445 non-null  object        
 8   product_class            19445 non-null  object        
 9   product_size             19445 non-null  object        
 10  list_price               19445 non-null  float64       
 11  standard_cost            19445 non-null  float64       
 12  product_first_sold_date  19445 n

-Finally the $ in the standard_cost column but not in the list_price column... for uniformity stick to one. make both currency datatype

### Assessing the NewCustomerList file

In [20]:
NewCustomerList_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [21]:
# checking for duplicates
NewCustomerList_df.duplicated().sum()

0

- No duplicate values

In [22]:
NewCustomerList_df.isna().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Unnamed: 16                              0
Unnamed: 17                              0
Unnamed: 18                              0
Unnamed: 19                              0
Unnamed: 20                              0
Rank                                     0
Value                                    0
dtype: int6

- The last name na values should be left untouched
- those with no stated dob should be dropped as they have gender value of U also
- 


In [23]:
#dropping na values in all but the last name column
NewCustomerList_df.dropna(subset=['DOB', 'job_title', 'job_industry_category'], inplace=True)
NewCustomerList_df.isna().sum()

first_name                              0
last_name                              20
gender                                  0
past_3_years_bike_related_purchases     0
DOB                                     0
job_title                               0
job_industry_category                   0
wealth_segment                          0
deceased_indicator                      0
owns_car                                0
tenure                                  0
address                                 0
postcode                                0
state                                   0
country                                 0
property_valuation                      0
Unnamed: 16                             0
Unnamed: 17                             0
Unnamed: 18                             0
Unnamed: 19                             0
Unnamed: 20                             0
Rank                                    0
Value                                   0
dtype: int64

In [24]:
#checking gender column values for U value
NewCustomerList_df.gender.value_counts()

Female    380
Male      355
Name: gender, dtype: int64

- deceased indicator is all null so should be removed as it is redundant
- gender U value is undefined and is taken care of by removing DOB null values
- The five unnamed columns are also unclear and should be clarified
- The country column should also be dropped as they are all in Australia
- The state abbreviations should be written in full for people like me who knoww not what they stand for
- property_valuation should take only integers


In [25]:
# dropping redundant and unclear columns
NewCustomerList_df.drop(['deceased_indicator', 'country', 'Unnamed: 20', 'Unnamed: 19', 'Unnamed: 18', 'Unnamed: 17', 'Unnamed: 16'],inplace= True, axis=1)
NewCustomerList_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 735 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           735 non-null    object        
 1   last_name                            715 non-null    object        
 2   gender                               735 non-null    object        
 3   past_3_years_bike_related_purchases  735 non-null    int64         
 4   DOB                                  735 non-null    datetime64[ns]
 5   job_title                            735 non-null    object        
 6   job_industry_category                735 non-null    object        
 7   wealth_segment                       735 non-null    object        
 8   owns_car                             735 non-null    object        
 9   tenure                               735 non-null    int64         
 10  address       

- post code and address are related too...so why still have address?

### Assessing the CustomerAddress file

In [26]:
CustomerAddress_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [27]:
# checking for duplicates
CustomerAddress_df.duplicated().sum()

0

In [28]:
# checking for null values
CustomerAddress_df.isna().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

- There are no duplicates and no null values

In [29]:
CustomerAddress_df.state.value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

- the names of the states should all be changed to abbreviatons

### Assessment of Customer Demographic sheet

In [30]:
CustomerDemographic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [31]:
CustomerDemographic_df.isna().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [32]:
CustomerDemographic_df.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

- gender column has same values in different text format. e.g F, Femal, Female... all for the female 
- null values in columns
- customer id 34 has a DOB of 1843 and is not deceased... this seems like an entry error
- default column is not well defined and contains nothing clear for now
- columns with no     do should be checked...also DOB should have constraints in filling


- I finally see a use for the deceased indicator column
- also the default column is unclear 
-

In [34]:
CustomerDemographic_df.customer_id.duplicated().sum()

0

In [36]:
Transactions_df.customer_id.duplicated().sum()

15953

In [51]:
CustomerAddress_df['customer_id'].countd()

AttributeError: 'Series' object has no attribute 'countd'

In [50]:
Transactions_df['customer_id'].count()

19445

In [52]:
Transactions_df['customer_id'].nunique()

3492

In [53]:
CustomerAddress_df['customer_id'].nunique()

3999

In [54]:
CustomerDemographic_df['customer_id'].nunique()

4000

<class 'pandas.core.frame.DataFrame'>
Int64Index: 735 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           735 non-null    object        
 1   last_name                            715 non-null    object        
 2   gender                               735 non-null    object        
 3   past_3_years_bike_related_purchases  735 non-null    int64         
 4   DOB                                  735 non-null    datetime64[ns]
 5   job_title                            735 non-null    object        
 6   job_industry_category                735 non-null    object        
 7   wealth_segment                       735 non-null    object        
 8   owns_car                             735 non-null    object        
 9   tenure                               735 non-null    int64         
 10  address       

### Exploratory Data Analysis
    This segment contains the details of task 2. 